src/main/java/com/exactpro/th2/crawler/Crawler.java

```diff
@@ -52,205 +52,238 @@
-                    report = sendMessages(crawlerId, info, batchSize);
-                } else {
-                    throw new ConfigurationException("Type must be either EVENTS or MESSAGES");
-                }
+            } else if (MESSAGES.equals(interval.getCrawlerType())) {
+                Map<String, RecoveryState.InnerMessage> lastProcessedMessages;
+                RecoveryState state = RecoveryState.getStateFromJson(interval.getRecoveryState());
 
-                restartPod = report.action == CrawlerAction.STOP;
+                lastProcessedMessages = state == null ? null : state.getLastProcessedMessages();
 
-                if (report.action == CrawlerAction.NONE) {
-                    interval = intervalsWorker.setIntervalProcessed(interval, true);
+                Map<String, MessageID> startIds = null;
 
-                    RecoveryState previousState = interval.getRecoveryState();
+                MessageID.Builder builder = MessageID.newBuilder();
 
-                    if (EVENTS.equals(interval.getCrawlerType())) {
-                        RecoveryState state = new RecoveryState(
-                                null,
-                                previousState.getLastProcessedMessages(),
-                                numberOfEvents,
-                                previousState.getLastNumberOfMessages());
+                if (lastProcessedMessages != null) {
+                    if (!fetchIntervalReport.processFromStart) {
+                        List<MessageID> ids = lastProcessedMessages.values().stream()
+                                .map(innerMessage -> {
+                                    com.exactpro.th2.common.grpc.Direction direction;
 
-                        interval = intervalsWorker.updateRecoveryState(interval, state);
-                    }
+                                    if (innerMessage.getDirection() == Direction.FIRST)
+                                        direction = com.exactpro.th2.common.grpc.Direction.FIRST;
+                                    else
+                                        direction = com.exactpro.th2.common.grpc.Direction.SECOND;
+
+                                    return builder
+                                            .setSequence(innerMessage.getSequence())
+                                            .setConnectionId(ConnectionID.newBuilder().setSessionAlias(innerMessage.getSessionAlias()).build())
+                                            .setDirection(direction)
+                                            .build();
+                                })
+                                .collect(Collectors.toList());
 
-                    numberOfEvents = 0L;
-                    numberOfMessages = 0L;
+                        startIds = ids.stream().collect(Collectors.toMap(messageID -> messageID.getConnectionId().getSessionAlias(),
+                                Function.identity(), (messageID, messageID2) -> messageID2));
```

Comment:
Why do you need the `ids` collection? If you don't want a long call chain you can store stream into a variable. And then call other operations you need (if you need to process stream only once). But in this case, the only operation left is a `collect(toMap())` so you don't even need to extract stream into a variable, in my opinion.